### PR TITLE
Implement mini-batch Sthocastic Gradient

### PR DESCRIPTION
**Pull request by RogerioOMDS**

### CODE REVIEW

File: src/sthocastic_gradient.jl [hunk 1]
```diff
@@ -39,36 +44,62 @@ function sthocastic_gradient(
...
+  end
 
+  γ = if learning_rate == :optimal
+    iter -> γ0 / (α * (1e3 + iter))
+  elseif learning_rate == :invscaling
+    iter -> γ0 / (iter + 1)^ν
+  elseif learning_rate == :time_based
+    iter -> γ0 / (1 + d * iter)
+  elseif learning_rate == :step_based
+    iter -> γ0 * d^(div(1 + iter, r))
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    iter -> γ0 * d ^ div(1 + iter, r)
```

File: src/sthocastic_gradient.jl [hunk 2]
```diff
@@ -39,36 +44,62 @@ function sthocastic_gradient(
...
+  
+  if learning_rate == :optimal && α == 0.0
+    @warn("Can't use learning_rate optimal with α = 0.0. Changing to learning_rate step_based.")
+    learning_rate = :step_based
+  end
 
+  γ = if learning_rate == :optimal
+    iter -> γ0 / (α * (1e3 + iter))
+  elseif learning_rate == :invscaling
+    iter -> γ0 / (iter + 1)^ν
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    iter -> γ0 / (iter + 1) ^ ν
```

File: src/sthocastic_gradient.jl [hunk 3]
```diff
@@ -14,21 +14,26 @@ function sthocastic_gradient(
...
+    γ0::Float64 = 1.0 # first step
   )
 
   iter = 0
   start_time = time()
   β = nlp.meta.x0
-  βavg = similar(β)
   n = nlp.meta.nobj
+  p = nlp.meta.nvar
+  βavg = similar(β)
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl:32]:
```suggestion
  βavg = similar(β)
  g_tmp = similar(β)
```

File: src/sthocastic_gradient.jl [hunk 4]
```diff
@@ -39,36 +44,62 @@ function sthocastic_gradient(
...
-      β -= γ * (α * P(β) + grad!(nlp, i, β, g))
+    Rindex = shuffle(1:n)
+    for l = 1:num_batches
+      index = if l < num_batches
+        batch_size * (l - 1) .+ (1:batch_size)
+      else
+        batch_size * (l - 1) + 1:n
+      end
+      
+      grad = zeros(p)
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
      g .= 0
```

File: src/sthocastic_gradient.jl [hunk 5]
```diff
@@ -39,36 +44,62 @@ function sthocastic_gradient(
...
+      for i in index
+        grad!(nlp, Rindex[i], β, g) 
+        grad += g
+      end
+      β -= γ(iter) * (α * P(β) + grad / length(index))
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
        grad!(nlp, Rindex[i], β, g_tmp)
        g += g_tmp
      end
      β -= γ(iter) * (α * P(β) + g / length(index))
```

**Review by abelsiqueira — APPROVED** [review]
I made some name changes. It's usually better to avoid using the same name as a function (`grad` is an NLPModels function).
Aside from that, I'm accepting.

**Comment by abelsiqueira** [comment]
Thanks
